### ISSUE TITLE
Dynamically change translations and translations without gettext

### ISSUE THREAD

**Issue by ogoffart** [slint-ui/slint#3715]
We need to support user changing translations through the interface, as well as translation on platform that do not have gettext (MCU)

The idea would be somehow to have API to provide a translator.

Perhaps we can re-use an interface similar to  https://docs.rs/tr/latest/tr/trait.Translator.html
and have a global setter to set it. That setter would mark all translated string as dirty.

( This should also help for https://github.com/slint-ui/slint/issues/3307 )

**Comment by qarmin** [comment]
I know that this has been discussed before and that the idea was already discarded, but isn't it worthwhile, however, to give users the option of using fluent-rs instead of gettext, since it works very well under any system, including windows.

This way, the external dependence on the code in C can be removed and it will make cross-compilation easier.

**Comment by ogoffart** [comment]
Regarding fluent, this can be done by using a global.
The old documentation should five hint https://slint.dev/releases/1.0.2/docs/slint/src/recipes/recipes#translations
So you would have in your code something like `text: Fluent.translate("my-text", ["FooBar"])`
Other possibilities are also possible: https://github.com/slint-ui/slint/issues/33#issuecomment-1275856180

**Comment by CarbonPool** [comment]
Any progress on changing the language dynamically?

**Comment by tronical** [comment]
Changing translations dynamically (when using gettext) is implemented in the master branch. This ticket was split and slightly confusing, so I've refocused it on the gettext removal.

**Comment by ogoffart** [comment]
I'm proposing to have the following:
 - `slint_build::CompilerConfiguration::bundle_translations(path, ...)`  in Rust 
 - `--bundle-translations <path>` in C++ (with corresponding API in our cmake interface)

The the compiler would look at the content of <path> and look for file with a given pattern.
Either:
 - `<language>.po`   (eg: `de.po`/`fr.po`)
 - `<language>/LC_MESSAGES/<domain_name>.po` (eg `de/LC_MESSAGES/gallery.po`) that's what we have in our example right now.
 - `<language>/LC_MESSAGES/<domain_name>.mo` we also have that. This is a binary format.
 - `<language>/<domain_name>.{mo,po}` drop the LC_MESSAGES ?

We can use crates like `gettext` (to open .mo) or `polib` (to open .po)
Then at compile time, we translate `@tr("Hello {}", name)` to something like (pseudo-code)
```rust
   private_api::format_string(match private_api::language() {
       1 => "Bonjour {}",
       2 => "Hallo {}",
       _ => "Hello {}",
   },&[name]);
```
Or maybe
```
   private_api::translation(&["Hello {}", "Bonjour {}", "Hallo {}"], &[name])
```

We can even have warning if a string or translated string don't have the right format, (or doesn't exist ?)

Now we still need a way to set or change the language.
I'm thinking there could be `slint::set_language(???)` but how does it map the language string to a number? Or we could generate a  `slint_set_language(&str)` function in the generated code. Or, we could have a slint function `set-language("fr")` in .slint

**Comment by tronical** [comment]
I agree that this is the way to go, roughly.

In terms of language selection: I think the build system API should explicitly say which languages/translations to pick. This permits creating multi-language builds as well as region-specific builds. Later the list of support translations could also go into the project file, to enable access in the live-preview.

Generally, I think it makes sense to see translations like resources/assets and bundle them, also on targets with operating systems, as opt-in.

**Comment by ubruhin** [comment]
I'm currently struggling with setting up translations for the new LibrePCB UI and I think a generic translators API as suggested in the initial post would be very helpful.

My situation:

- A large C++ backend with many strings, using Qt's translation toolkit and a whole workflow set up (Transifex, deployment, dynamic loading at runtime).
- The new Slint UI contains a lot of translatable strings too, using `@tr()`.

For the C++ backend I want to keep the current system - it would be way too much effort to change it, and actually Qt provides a mature translations toolkit which works very well and flawlessly across all platforms.

For the Slint UI there are two solutions provided:
- Gettext: I gave it a try, but honestly this thing feels everything else than flawlessly. I could not even manage to compile it on Windows within reasonable time such that it can be used from my CMake project. The build system looks like 30 years old and not portable at all. Not really the kind of dependency I want for LibrePCB.
- Bundled translations: Would be very strange to have UI translations compiled into the application while backend strings are loaded at runtime. Also after a quick sight I do not see the possibility to specify multiple languages as fallbacks (e.g. using "de_CH" as language, but fall back to "de" for string not existing in "de_CH").

In either way, in the end I prefer to use the same system for both the UI and the backend, to keep a common workflow, consistent behavior, and low maintenance effort. So I wonder if I could use Qt's mature and painless translation system for the Slint UI?

One option would be to use a custom callback (implemented in the backend) instead of `@tr()`. However, I don't see an elegant way to pass the translation context (`msgctxt`) without needing to specify it manually in every call. Also a custom callback is way less convenient than `@tr()` (no optional arguments etc.).

So I wonder if I could use `@tr()` and somehow register my own translation function in the backend? This would also allow to use `slint-tr-extractor` for the string extraction - the *.po can then easily be converted to Qt's *.ts file format (or any other).

One tricky thing however would be placeholders handling. Slint uses `{}` and `{n}` while Qt uses `%1` and `%n`. I'm not sure if platforms like Transifex will handle Qt translation files correctly if Slint's syntax is used. Probably I'd rather convert the *.po with Slint syntax to *.ts with Qt syntax, and convert it back in the C++ backend. But this would work only if the callback for `@tr()` lets me do the placeholder substitution on my own.

Would such a custom `@tr()` handler be reasonable solution, and maybe even a low-hanging fruit?

Any tip would be much appreciated.

**Comment by ubruhin** [comment]
I just saw `gettext-rs` even doesn't have CI for Windows: https://github.com/gettext-rs/gettext-rs/tree/master/.github/workflows

So it's not really a surprise that this crate causes troubles on Windows :-/

Now I investigated a bit the approach of a translate callback. It looks like [this call](https://github.com/slint-ui/slint/blob/5766c22ecbde64a8857a2659cad7c8afccc5b4ed/internal/core/translations.rs#L195) from `translate()` to `translate_gettext()` is exactly the place where Qt's (or any other toolkit's) translate function could be invoked. I implemented a PoC where `translate_gettext()` is replaced by a callback into my C++ backend and so far this seems to work perfectly fine.

**Comment by ubruhin** [comment]
I see in 2bbe5e87867a4efbfcbcf26b7b1f76e0d19f3125 there was a function `SlintContext.set_external_translator()` added, would this allow using a custom translator now (e.g. a callback into our C++ code base)? 🤔

**Comment by tronical** [comment]
Yes, exactly. We just need a C++ API that wraps the Rust trait (base class with virtuals that forwards to a Rust impl of the trait)

**Comment by ubruhin** [comment]
I made a proposal for registering an external translator from C++: https://github.com/slint-ui/slint/pull/10979
It's open for discussion about the concept :)

**Comment by castarco** [comment]
Anyone around who can explain a bit to newbies what is the state of this issue and what are the blockers? I assume that compatibility with other languages such as C++ (I come from Rust) makes it a bit more complicated than if it was for just Rust.

On my side, I've been using `i18n-embed` and `i18n-embed-fl` for my translations on other projects without Slint, but I'm struggling to see how to use that same (or similar) setup for my new Slint codebase... basically because many strings are set inside the `.slint` files, and I don't have as much freedom inside these files as I have in pure Rust code.

I'd say "Fluent" translations seem much simpler to manage from a developer perspective (although I'm managing them "manually", so I'm not so sure about the tooling for this format).

**Comment by ogoffart** [comment]
If you want fluent-like translation with String outside of the .slint file, see
https://github.com/slint-ui/slint/issues/3715#issuecomment-1798221602
